compiler/rustc_const_eval/src/const_eval/error.rs

```diff
@@ -162,14 +163,15 @@ where
             // Report as hard error.
             let (our_span, frames) = get_span_and_frames();
             let span = span.unwrap_or(our_span);
-            let err = mk(span, frames);
-            let mut err = tcx.sess.create_err(err);
 
-            let msg = error.diagnostic_message();
-            error.add_args(&tcx.sess.parse_sess.span_diagnostic, &mut err);
+            let handler = &tcx.sess.parse_sess.span_diagnostic;
+            let mut err = handler.create_err(Spanned { span, node: InterpErrorExt(error) });
+
+            for frame in frames {
+                err.eager_subdiagnostic(handler, frame);
+            }
 
             // Use *our* span to label the interp error
-            err.span_label(our_span, msg);
```

Comment:
So this part is probably why the ui tests are now different. To fix this you probably need to eagerly translate the diagnostic message to string and add a label.

Reply:
Yeah that fixed the majority of the ui tests. Although now some are failing because the help messages are missing.

Now this makes me wonder. If we need to add this to a span message, shouldn't the newtypes better implement the `rustc_errors::AddToDiagnostic`?

```rust
impl AddToDiagnostic for InterpErrorExt {
    fn add_to_diagnostic_with<F>(self, diag: &mut rustc_errors::Diagnostic, f: F)
    where
        F: Fn(
            &mut rustc_errors::Diagnostic,
            rustc_errors::SubdiagnosticMessage,
        ) -> rustc_errors::SubdiagnosticMessage,
    {
        diag.set_arg("named", self.named);
        let msg = f(diag, crate::fluent_generated::builtin_macros_format_unused_arg.into());

        // here we could add our span labels with our message
        diag.span_label(self.span, msg);

        // here we can add our help messages
        diag.help("some help message");
    }
}
```

Reply:
I don't think that is possible. Helps and labels have their own variables and they should be eagerly substituted to prevent any pollution with named arguments. Having the `IntoDiagnostic` impl and doing the rest by hand might be the best we can do. But you should feel free to experiment whether `AddToDiagnostic` works.

Reply:
Actually though, it might be possible. It's just that we have to set the primary message of the diagnostic and the label by eagarly substituting.

Reply:
I'll try to experiment implementing `AddToDiagnostic`.